### ISSUE TITLE
@ViewChildren QueryList has broken interaction with Change Detection

### ISSUE THREAD

**Issue by alxhub** [angular/angular#10962]
`@ViewChildren`'s `QueryList` interacts in a bad manner with change detection, which results in exceptions in dev mode and broken behavior in prod mode.

The fundamental problem is that the `QueryList` is updated after change detection has occurred. This causes two error-prone effects:

1) Bindings to the `QueryList`, including `QueryList.length`, take on new values after change detection, causing them to throw change-after-checked errors in dev mode.

Reproduction/example: http://plnkr.co/edit/c3QHO1MYlLbd3op8eC1E

2) The `QueryList.changes` `Observable` has its subscribers run after change detection, meaning that anything they do will not be change detected, and changes to the template cause change-after-checked errors.

Reproduction/example: http://plnkr.co/edit/YBRHtSrHJ4gzcVBLubNz

These problems seem fairly serious, as anyone who attempts to use `@ViewChildren` in an app will encounter them.

**Comment by mhevery** [comment]
I think this is working as intended. In order for us to compute the `QueryList` we have to run the change detection. This is because during the change detection we could unroll `ngFor` and `ngIf` which may change the content of `QueryList`.

**Comment by pkozlowski-opensource** [comment]
I also was under the impression that it works as designed - at least I was living with assumptions pointed out by @mhevery. For me it makes since querying view children means that the view was constructed already which implies that change detection run. As such a person should assume that no data model can be touched in `ngAfterViewInit` / `ngAfterViewChecked`. Indeed I was assuming that those life-cycle hooks are here to do last-minute DOM adjustments (ex.  set focus or sth) and definitively not to manipulate model. 

Well, at least this is how I was thinking about this life-cycle part until today :-)

**Comment by alxhub** [comment]
I agree that this is mostly a documentation problem. In some sense, the very concept of a component observing its own view is similar to two way binding, since a change in the model drives a change in the view which then could (theoretically) change the model again.

**Comment by naomiblack** [comment]
@wardbell could you figure out a good place to cover in our docs?

**Comment by frickt** [comment]
@mhevery @pkozlowski-opensource till rc.5 it was possible to use `ComponentResolver `in `ngAfterViewInit`-hook to create dynamic content depending on the `ViewChildren` `QueryList `elements, now this causes change-after-checked error. Was that intended? If yes, is there an alternative?

**Comment by wardbell** [comment]
`QueryList` isn't documented in the guides yet and is not on the short list for coverage soon.

Seems to me that this point should be noted in the API docs as well.

**Comment by jenniferfell** [comment]
It looks like this has been resolved. 

https://angular.io/api/core/QueryList
https://angular.io/api/core/ViewChildren
https://angular.io/api/core/ContentChildren
https://angular.io/guide/dependency-injection-in-action

"You can use ViewChildren to get the QueryList of elements or directives from the view DOM. Any time a child element is added, removed, or moved, the query list will be updated, and the changes observable of the query list will emit a new value. View queries are set before the ngAfterViewInit callback is called."

@alxhub May I close this as resolved?
